eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/InvoicePurger.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2022 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment.receive
+
+import akka.Done
+import akka.actor.typed.Behavior
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import fr.acinq.eclair.db.PaymentsDb
+import fr.acinq.eclair.payment.receive.InvoicePurger.{Command, PurgeCompleted, PurgeResult, TickPurge}
+import fr.acinq.eclair.{TimestampMilli, TimestampMilliLong}
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success, Try}
+
+/**
+ * This actor will purge expired invoices from the database it was initialized with at a scheduled interval.
+ */
+object InvoicePurger {
+
+  sealed trait Command
+  private case object TickPurge extends Command
+  private case class PurgeResult(result: Try[Done]) extends Command
+
+  sealed trait PurgeEvent
+  // this notification is sent when we have completed our invoice purge process
+  case object PurgeCompleted extends PurgeEvent
+
+  def apply(paymentsDb: PaymentsDb, interval: FiniteDuration): Behavior[Command] =
+    Behaviors.setup { context =>
+      // wait for purge events sent at `interval`
+      Behaviors.withTimers { timers =>
+        timers.startTimerAtFixedRate(TickPurge, interval)
+        new InvoicePurger(paymentsDb, context).waiting()
+      }
+    }
+}
+
+class InvoicePurger private(paymentsDb: PaymentsDb,
+                            context: ActorContext[Command]) {
+
+  // purge at each tick unless currently purging
+  def waiting(): Behavior[Command] =
+    Behaviors.receiveMessagePartial {
+      case TickPurge =>
+        doPurge()
+        purging()
+    }
+
+  // ignore ticks and wait for for purge to complete
+  def purging(): Behavior[Command] =
+    Behaviors.receiveMessagePartial {
+      case PurgeResult(res) =>
+        res match {
+          case Success(Done) => context.log.debug("invoice purge succeeded")
+          case Failure(cause) => context.log.warn(s"invoice purge failed: $cause")
+        }
+        waiting()
+    }
+
+  private def doPurge(): Unit = {
+    val now = TimestampMilli.now()
+    val expiredPayments = paymentsDb.listExpiredIncomingPayments(0 unixms, now)
```

Comment:
It's a bit inefficient to always scan *all* invoices (`from = 0 unixms`). If your node has received a lot of payments (that have been correctly paid), that's a lot of DB rows that we will re-scan regularly for no reason. What we could do to improve this without over-complicating it would be to scan the whole DB on the first run (when eclair has been restarted) and then only scan recent rows on the following runs.

A simple way to do that would be:

- add a boolean to the `waiting` state: `def waiting(fullScan: Boolean): Behavior[Command]`
- start with `fullScan = true`, and then transition to `fullScan = false` after the first run

It would look something like:

```scala
  def waiting(fullScan: Boolean): Behavior[Command] =
    Behaviors.receiveMessage {
      case TickPurge =>
        val now = TimestampMilli.now()
        val start = if (fullScan) 0 unixms else now - 15.days
        // purge expired payments    
        val expiredPayments = paymentsDb.listExpiredIncomingPayments(start, now)
        expiredPayments.foreach(p => paymentsDb.removeIncomingPayment(p.invoice.paymentHash))

        // publish a notification when we have purged expired invoices
        if (expiredPayments.nonEmpty) {
          context.system.eventStream ! EventStream.Publish(PurgeCompleted)
        }
        
        waiting(fullScan = false)
    }
```

The trick is that the interval chosen (15 days) must be greater than the invoice expiry. 15 days feels like a good trade-off, invoices should very likely expire sooner than that, and even if some invoices use a longer expiry, they will be cleaned up at the next full scan (when eclair restarts).